Refactor

ADmad opened the pull request:
- A separate cache key is no longer used to store the limit reset timestamp.
Instead the limit, remaining request and limit reset timestamp are stored under
same key. On each valid request within the allowed period the TTL of rate
limiting cache is also properly decremented.
- Allow modifying throttling info using callback. This effectively allows changing the rate limit for requests
based on any desired criteria.

ADmad committed:
Refactor the ThrottleMiddleware.

A separate cache key is no longer used to store the limit reset timestamp.
Instead the limit, remaining request and limit reset timestamp are stored under
same key. On each valid request within the allowed period the TTL of rate
limiting cache is also properly decremented.

ADmad committed:
Allow modifying throttling info using callback.

This effectively allows changing the rate limit for requests
based on any desired criteria.

ADmad committed:
Include "Retry-After" header with error response.

smarek commented:
Should this override the #34 or they will coexist?

ADmad commented:
After this implementation is your request weight feature necessary? As per your example you can simply return a different key and limit for POST request to your login/registration pages.

smarek commented:
I'm thinking, it might have the same outcome, two things I'd like to see

- Progressive throttling, ie. first time you exceed the limit, cooldown period is 60, second time it's 120, third time it might be 3600 or completely different
- Configuration of throttling in routes, not in middleware configuration, eg. in my case there are like 350 URLs that now have to be handled by whole different classes/impl structure (callback), even if the configuration could be put in place when calling `RouteBuilder->connect`

I'm not sure even the first thing is now provided with #38 or #34 , if you want to do the whole thing right, could you please think about these use-cases?

ADmad commented:
> Progressive throttling, ie. first time you exceed the limit, cooldown period is 60, second time it's 120, third time it might be 3600 or completely different

Hmm.. Let's say you are allowed 5 requests / 10 seconds. Currently if you exceed your limit in 6 seconds then the cool down period is 4 seconds and if you exceed your limit in 9 seconds then cool down period is just 1 second. 

So to implement an increasing cool down period the "reset" timestamp and cache's TTL would have to be increased for every successive failed request. Instead of storing the "remaining" count I could store the "calls" count which would keep incrementing for every request. Then using a callback I allow modifying the reset timestamp and cache TTL.

> Configuration of throttling in routes, not in middleware configuration, eg. in my case there are like 350 URLs that now have to be handled by whole different classes/impl structure (callback), even if the configuration could be put in place when calling RouteBuilder->connect

You can used route scoped middlewares: https://book.cakephp.org/4/en/development/routing.html#connecting-scoped-middleware

ADmad committed:
Store number of calls instead of remaining count in cache.

This opens the possibility of increasing the cool down period for repeated
requests after limit is exhausted.

ADmad committed:
Use DTOs instead of arrays.

ADmad committed:
Fix typing errors

ADmad committed:
Rename DTO to value object.

smarek commented:
@ADmad 
1. agree, if there was a callback to provide modified reset timestamp and/or ttl, that would allow for progressive throttling
2. routing scoped middleware is not the same, simply put, i'd like to configure the throttling params, same as you shown in symfony's RateLimitBundle, however since annotations based solution is not ideal currently, i'd like to see something like this

**simple example without callback**
```php
// routes.php
$routes->connect(
  '/sign-in',
  ['controller' => 'Users', 'action' => 'login'],
  ['_name' => 'login', '_throttle_config' => [
    [
      'method' => ['PUT', 'POST'],
      'period' => 3600,
      'limit' => 10
    ],
    [
      'method' => 'GET',
      'period' => 120,
      'limit' => 10
    ]
  ]]
);
```

**simple example using callback / providing implementation**
```php
// routes.php
$routes->connect(
    '/sign-in',
    ['controller' => 'Users', 'action' => 'login'],
    // better would be class and method, but for start class implementing expected interface will suffice
    ['_name' => 'login', '_throttle_callback' => App\Throttle\LoginThrottle::class]
);
```

However I'm not sure, it's currently possible to provide additional config params through standard `RouteBuilder`, so i'm thinking of subclass such as `ThrottleRouteBuilder` and if the throttle configuration of route is provided, Builder would automatically configure the middleware through some shared config or alike?

ADmad committed:
Use events instead of config callbacks.

ADmad commented:
I have switched from using config options for callbacks to using events. You now have events which allow modifying the identifier, throttling info and rate limiting info before it's cached.

So you can explore making your own routes builder and setup listener for required events to control various aspects of rate limiting.

ADmad committed:
Update event name

ADmad committed:
Add back ability to set callback for identifier through config

ADmad committed:
Rename event

ADmad committed:
Use config key instead of static property for cache config name.

ADmad committed:
Add exception for removed config

ADmad committed:
Removed defunct static var

ADmad committed:
Update readme

ADmad committed:
Show warning instead of throwing exception
